cmd/kubeadm/app/phases/uploadconfig/uploadconfig_test.go

```diff
@@ -19,37 +19,79 @@ package uploadconfig
 import (
 	"testing"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 	clientsetfake "k8s.io/client-go/kubernetes/fake"
+	core "k8s.io/client-go/testing"
 	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
 )
 
 func TestUploadConfiguration(t *testing.T) {
 	tests := []struct {
-		name    string
-		cfg     *kubeadmapi.MasterConfiguration
-		wantErr bool
+		name           string
+		createErr      error
+		updateErr      error
+		idempotentTest bool
+		wantErr        bool
+		verifyResult   bool
 	}{
 		{
-			"basic validation with correct key",
-			&kubeadmapi.MasterConfiguration{
-				KubernetesVersion: "1.7.3",
-			},
-			false,
+			name:         "basic validation with correct key",
+			wantErr:      false,
+			verifyResult: true,
+		},
+		{
+			name:           "update existing should report no error",
+			idempotentTest: true,
+			wantErr:        false,
+			verifyResult:   true,
+		},
+		{
+			name:      "unexpected errors for create should be returned",
+			createErr: apierrors.NewUnauthorized(""),
+			wantErr:   true,
+		},
+		{
+			name:           "unexpected errors for update should be returned",
+			updateErr:      apierrors.NewUnauthorized(""),
+			idempotentTest: true,
+			wantErr:        true,
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			cfg := &kubeadmapi.MasterConfiguration{
+				KubernetesVersion: "1.7.3",
+			}
 			client := clientsetfake.NewSimpleClientset()
-			if err := UploadConfiguration(tt.cfg, client); (err != nil) != tt.wantErr {
+			if tt.createErr != nil {
+				client.PrependReactor("create", "configmaps", func(action core.Action) (bool, runtime.Object, error) {
+					return true, nil, tt.createErr
+				})
+			}
+			// For idempotent test, we check the result of the second call.
+			if err := UploadConfiguration(cfg, client); !tt.idempotentTest && ((err != nil) != tt.wantErr) {
```

Comment:
I'm finding it difficult to follow these conditionals to figure out what is really going on, and when `idempotentTest` should or shouldn't be set. Do we really need it? It seems like we can represent the tests based only on what errors should be returned during `client.create()` / `client.update()` and what error we expect `UploadConfiguration` to return under those conditions.

Reply:
idempotentTest is set only when we call UploadConfiguration twice. I will update comments